# Семинар 1: знакомство с NumPy


## Немного про Jupyter notebook

Полная документация: https://devpractice.ru/python-lesson-6-work-in-jupyter-notebook/

---
В Jupyter Notebook есть два режима работы: режим _команд_ и режим _редактирования_

_Командный_ режим нужен для того, чтобы взаимодействовать и управлять ячейками (добавлять, удалять, запускать, копировать, ...)

В режиме _редактирования_ вы меняете содержимое ячейки. 

Ячейки бывают двух основных типов, _код_ и _разметка_

### Полезные команды

(находясь в командном режиме)

- `a` - добавить пустую ячейку сверху

- `b` - добавить пустую ячейку снизу
- `c` - скопировать текущую ячейку
- `v` - вставить скопированную ячейку
- `d` - удалить текущую ячейку
- `x` - вырезать (удалить и скопировать) текущую ячейку
- `m` - изменить тип выбранной ячейки на "разметка" 
- `y` - изменить тип выбранной ячейки на "код" 
- `z` - отменить последнее действие


- `Enter` - начать редактировать выбранную ячейку

(будучи в режим редактирования ячейки)
- `esc` - вернуться в командный режим

(будучи в любом режиме)

- `Ctrl + Enter` - запустить выбранную ячейку
- `Shift + Enter` - запустить выбранную ячейку и выбрать следующую

In [6]:
2 + 2

4

## NumPy

**NumPy** — библиотека языка Python, позволяющая (удобно) работать с многомерными массивами и матрицами. Кроме того, NumPy позволяет векторизовать многие вычисления, имеющие место в машинном обучении.

 - [numpy](http://www.numpy.org)
 - [numpy tutorial](http://cs231n.github.io/python-numpy-tutorial/)
 - [100 numpy exercises](http://www.labri.fr/perso/nrougier/teaching/numpy.100/)
 
Кстати, про  `NumPy` недавно вышла [публикация](https://www.nature.com/articles/s41586-020-2649-2) в Nature.

Библиотека numpy является удобным инструментом для работы с многомерными массивами с возможностью векторизации вычислений. Рассмотрим базовые вещи, которые можно делать с помощью нее.

In [7]:
import numpy as np
import warnings
warnings.filterwarnings('ignore')

Основным типом данных NumPy является многомерный массив элементов одного типа — [numpy.ndarray](http://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.array.html). Каждый подобный массив имеет несколько *измерений* или *осей* — в частности, вектор (в классическом понимании) является одномерным массивом и имеет 1 ось, матрица является двумерным массивом и имеет 2 оси и т.д.

In [ ]:
vec = np.array([1, 2, 3])
vec.ndim # количество осей

: 

In [9]:
mat = np.array([[1, 2, 3], [4, 5, 6]])
mat.ndim

2

Чтобы узнать длину массива по каждой из осей, можно воспользоваться атрибутом shape:

In [10]:
vec.shape

(3,)

Чтобы узнать тип элементов и их размер в байтах:

In [11]:
mat.dtype.name

'int64'

In [12]:
mat.itemsize

8

#### Создание массивов

Есть несколько способов сформировать массив в NumPy:

* Передать итерируемый объект в качестве параметра функции array (можно также явно указать тип элементов):

In [13]:
A = np.array([1, 2, 3])
A, A.dtype

(array([1, 2, 3]), dtype('int64'))

In [14]:
A = np.array([1, 2, 3], dtype=float)
A, A.dtype

(array([1., 2., 3.]), dtype('float64'))

* Воспользоваться функциями zeros, ones, empty, identity, если вам нужен объект специального вида:

In [15]:
np.zeros((3,))

array([0., 0., 0.])

In [16]:
np.ones((3, 4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [17]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

* Воспользоваться функциями arange (в качестве параметров принимает левую и правую границы последовательности и **шаг**) и linspace (принимает левую и правую границы и **количество элементов**) для формирования последовательностей:

In [18]:
np.arange(2, 20, 3) # аналогично стандартной функции range python, правая граница не включается

array([ 2,  5,  8, 11, 14, 17])

In [19]:
np.arange(2.5, 8.7, 0.9) # но может работать и с вещественными числами

array([2.5, 3.4, 4.3, 5.2, 6.1, 7. , 7.9])

In [20]:
np.linspace(2, 18, 14) # правая граница включается (по умолчанию)

array([ 2.        ,  3.23076923,  4.46153846,  5.69230769,  6.92307692,
        8.15384615,  9.38461538, 10.61538462, 11.84615385, 13.07692308,
       14.30769231, 15.53846154, 16.76923077, 18.        ])

* Изменить размеры существующего массива с помощью reshape (при этом количество элементов должно оставаться неизменным):

In [21]:
np.arange(9).reshape(3, 3)

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

Вместо значения длины массива по одному из измерений можно указать -1 — в этом случае значение будет рассчитано автоматически:

In [22]:
np.arange(8).reshape(2, -1)

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

* Транспонировать существующий массив:

In [23]:
C = np.arange(6).reshape(2, -1)
C

array([[0, 1, 2],
       [3, 4, 5]])

In [24]:
C.T

array([[0, 3],
       [1, 4],
       [2, 5]])

**Повторить существующий массив:**
- Первое число (2) → дублируем массив по вертикали (по строкам) 2 раза
- Второе число (2) → дублируем массив по горизонтали (по столбцам) 2 раза

In [25]:
a = np.arange(3)
np.tile(a, (2, 2))

array([[0, 1, 2, 0, 1, 2],
       [0, 1, 2, 0, 1, 2]])

In [26]:
np.tile(a, (4, 1))

array([[0, 1, 2],
       [0, 1, 2],
       [0, 1, 2],
       [0, 1, 2]])

#### Базовые операции

* Базовые арифметические операции над массивами выполняются поэлементно:

In [27]:
A = np.arange(9).reshape(3, 3)
B = np.arange(1, 10).reshape(3, 3)

In [28]:
print(A)
print(B)

[[0 1 2]
 [3 4 5]
 [6 7 8]]
[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [29]:
A + B

array([[ 1,  3,  5],
       [ 7,  9, 11],
       [13, 15, 17]])

In [30]:
A * 1.0 / B

array([[0.        , 0.5       , 0.66666667],
       [0.75      , 0.8       , 0.83333333],
       [0.85714286, 0.875     , 0.88888889]])

In [31]:
A + 1

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [32]:
3 * A

array([[ 0,  3,  6],
       [ 9, 12, 15],
       [18, 21, 24]])

In [33]:
A ** 2

array([[ 0,  1,  4],
       [ 9, 16, 25],
       [36, 49, 64]])

Отдельно обратим внимание на то, что умножение массивов также является **поэлементным**, а не матричным:

In [34]:
A * B

array([[ 0,  2,  6],
       [12, 20, 30],
       [42, 56, 72]])

Для выполнения матричного умножения необходимо использовать функцию dot:

In [35]:
A.dot(B)

array([[ 18,  21,  24],
       [ 54,  66,  78],
       [ 90, 111, 132]])

Для умножения векторов или матриц можно также использовать оператор `@`:

In [36]:
A @ B

array([[ 18,  21,  24],
       [ 54,  66,  78],
       [ 90, 111, 132]])

In [37]:
np.array([1, 2, 3, 4]) @ np.array([1, 1, 1, 1])

10

Поскольку операции выполняются поэлементно, операнды бинарных операций должны иметь одинаковый размер. Тем не менее, операция может быть корректно выполнена, если размеры операндов таковы, что они могут быть расширены до одинаковых размеров. Данная возможность называется [broadcasting](http://www.scipy-lectures.org/intro/numpy/operations.html#broadcasting):
![](https://jakevdp.github.io/PythonDataScienceHandbook/figures/02.05-broadcasting.png)

In [39]:
A

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [38]:
np.tile(np.arange(0, 40, 10), (3, 1)).T + np.array([0, 1, 2])

array([[ 0,  1,  2],
       [10, 11, 12],
       [20, 21, 22],
       [30, 31, 32]])

В задачах, например, в обработке изображений, часто приходится работать с изображениями (массивами), где требуется применить некоторую операцию ко всем пикселям. Broadcast позволяет делать это быстрее, чем если бы вы использовали циклы для каждого пикселя.


Пример: Предположим, у вас есть изображение, и вы хотите увеличить яркость на постоянное значение для всех пикселей:

In [3]:
import numpy as np

image = np.array([[100, 150, 200], [50, 60, 70]])
brightness_increase = np.array([10, 20, 30])

new_image = image + brightness_increase  # Broadcasting
print(new_image)  #  Массив brightness_increase растягивается вдоль строк изображения, и прибавляется к каждому пикселю.

[[110 170 230]
 [ 60  80 100]]


* Некоторые операции над массивами (например, вычисления минимума, максимума, суммы элементов) выполняются над всеми элементами вне зависимости от формы массива, однако при указании оси выполняются вдоль нее (например, для нахождения максимума каждой строки или каждого столбца):

In [ ]:
A.min()

: 

In [41]:
A.max(axis=1)

array([2, 5, 8])

In [42]:
A.sum(axis=1)

array([ 3, 12, 21])

#### Индексация

Для доступа к элементам может использоваться [много различных способов](http://docs.scipy.org/doc/numpy/reference/arrays.indexing.html), рассмотрим основные.

* Для индексации могут использоваться конкретные значения индексов и срезы (slice), как и в стандартных типах Python. Для многомерных массивов индексы для различных осей разделяются запятой. Если для многомерного массива указаны индексы не для всех измерений, недостающие заполняются полным срезом (:).

In [43]:
a = np.arange(10)
a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [44]:
a[2:5]

array([2, 3, 4])

In [45]:
a[3:8:2]

array([3, 5, 7])

In [46]:
A = np.arange(81).reshape(9, -1)
A

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23, 24, 25, 26],
       [27, 28, 29, 30, 31, 32, 33, 34, 35],
       [36, 37, 38, 39, 40, 41, 42, 43, 44],
       [45, 46, 47, 48, 49, 50, 51, 52, 53],
       [54, 55, 56, 57, 58, 59, 60, 61, 62],
       [63, 64, 65, 66, 67, 68, 69, 70, 71],
       [72, 73, 74, 75, 76, 77, 78, 79, 80]])

In [47]:
A[2:4]

array([[18, 19, 20, 21, 22, 23, 24, 25, 26],
       [27, 28, 29, 30, 31, 32, 33, 34, 35]])

In [48]:
A[:, 2:4]

array([[ 2,  3],
       [11, 12],
       [20, 21],
       [29, 30],
       [38, 39],
       [47, 48],
       [56, 57],
       [65, 66],
       [74, 75]])

In [49]:
A[2:4, 2:4]

array([[20, 21],
       [29, 30]])

In [50]:
A[-1]

array([72, 73, 74, 75, 76, 77, 78, 79, 80])

* Также может использоваться индексация при помощи списков индексов (по каждой из осей):

In [51]:
A = np.arange(81).reshape(9, -1)
A

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23, 24, 25, 26],
       [27, 28, 29, 30, 31, 32, 33, 34, 35],
       [36, 37, 38, 39, 40, 41, 42, 43, 44],
       [45, 46, 47, 48, 49, 50, 51, 52, 53],
       [54, 55, 56, 57, 58, 59, 60, 61, 62],
       [63, 64, 65, 66, 67, 68, 69, 70, 71],
       [72, 73, 74, 75, 76, 77, 78, 79, 80]])

In [52]:
A[[2, 4, 5], [0, 1, 3]]

array([18, 37, 48])

* Может применяться логическая индексация (при помощи логических массивов):

In [53]:
A = np.arange(11)
A

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [54]:
A[A % 5 != 3]

array([ 0,  1,  2,  4,  5,  6,  7,  9, 10])

In [55]:
A[np.logical_and(A != 7, A % 5 != 3)] # также можно использовать логические операции

array([ 0,  1,  2,  4,  5,  6,  9, 10])

#### Зачем?

Зачем необходимо использовать NumPy, если существуют стандартные списки/кортежи и циклы?

Причина заключается в скорости работы. Попробуем посчитать скалярное произведение 2 больших векторов:

In [4]:
SIZE = 10000000

A_quick_arr = np.random.normal(size = (SIZE,))
B_quick_arr = np.random.normal(size = (SIZE,))

A_slow_list, B_slow_list = list(A_quick_arr), list(B_quick_arr)

In [5]:
%%time
ans = 0
for i in range(len(A_slow_list)):
    ans += A_slow_list[i] * B_slow_list[i]

CPU times: user 1.46 s, sys: 0 ns, total: 1.46 s
Wall time: 1.46 s


In [6]:
%%time
ans = sum([A_slow_list[i] * B_slow_list[i] for i in range(SIZE)])

CPU times: user 1.03 s, sys: 66.9 ms, total: 1.1 s
Wall time: 1.1 s


In [9]:
%%time
ans = np.sum(A_quick_arr * B_quick_arr)

CPU times: user 20.3 ms, sys: 12.3 ms, total: 32.6 ms
Wall time: 31.4 ms


In [10]:
%%time
ans = A_quick_arr.dot(B_quick_arr)

CPU times: user 14.3 ms, sys: 16.3 ms, total: 30.6 ms
Wall time: 4.28 ms


NumPy работает быстро по нескольким причинам:
* Массивы хранятся в непрерывном участке памяти, а все элементы имеют один и тот же тип
* Для вычислений по возможности используются библиотеки линейной алгебры вроде BLAS  (Basic Linear Algebra Subprograms): Это библиотека, предоставляющая базовые операции линейной алгебры (например, умножение матриц)

Посмотреть, какая библиотека используется у вас, можно в конфигурации NumPy:

In [11]:
print(np.show_config())

Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /home/nadine/.config/jupyterlab-desktop/jlab_server/include
    lib directory: /home/nadine/.config/jupyterlab-desktop/jlab_server/lib
    name: blas
    openblas configuration: unknown
    pc file directory: /home/nadine/.config/jupyterlab-desktop/jlab_server/lib/pkgconfig
    version: 3.9.0
  lapack:
    detection method: internal
    found: true
    include directory: unknown
    lib directory: unknown
    name: dep140023363602336
    openblas configuration: unknown
    pc file directory: unknown
    version: 1.26.4
Compilers:
  c:
    args: -march=nocona, -mtune=haswell, -ftree-vectorize, -fPIC, -fstack-protector-strong,
      -fno-plt, -O2, -ffunction-sections, -pipe, -isystem, /home/nadine/.config/jupyterlab-desktop/jlab_server/include,
      -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/numpy_1707225359967/work=/usr/local/src/conda/numpy-1.26.4,
      -fdebug-prefix

### Примеры векторизации вычислений на NumPy


Разберём несколько задач (из [100 numpy exercises](http://www.labri.fr/perso/nrougier/teaching/numpy.100/)), где NumPy может существенно ускорить вычисления и упростить код.

##### Дан четырёхмерный массив. Как получить двумерный массив, в котором элемент с индексами $(i, j)$ содержит сумму всех элементов исходного массива, у которых первые два индекса — это $(i, j)$?

In [56]:
A = np.random.randint(0,1000,(2,5,20,25))
res = A.reshape(A.shape[:-2] + (-1,)).sum(axis=-1)
print(res)

[[258340 251179 238200 252220 246860]
 [243073 245739 251546 249450 246257]]


Создаем четырёхмерный массив A размером (2, 5, 20, 25), где:

Размерность 0 (индексы от 0 до 1) — это 2.
Размерность 1 (индексы от 0 до 4) — это 5.
Размерность 2 (индексы от 0 до 19) — это 20.
Размерность 3 (индексы от 0 до 24) — это 25.



Задача, которую вы описываете, сводится к агрегации значений из четырёхмерного массива с суммированием по определённой оси, чтобы получить двумерный массив.

Давайте разберём, что происходит в данном коде:

### Исходный массив:
```python
A = np.random.randint(0, 1000, (2, 5, 20, 25))
```
Это создаёт четырёхмерный массив **A** размером (2, 5, 20, 25), где:
- Размерность 0 (индексы от 0 до 1) — это 2.
- Размерность 1 (индексы от 0 до 4) — это 5.
- Размерность 2 (индексы от 0 до 19) — это 20.
- Размерность 3 (индексы от 0 до 24) — это 25.

### Основная цель:
Нам нужно получить двумерный массив, где каждый элемент на позиции \([i, j]\) будет содержать сумму всех элементов исходного массива, для которых первые два индекса равны \(i\) и \(j\) соответственно. То есть, нужно суммировать все элементы с фиксированными первыми двумя индексами (i и j) и объединить их в двумерный массив.


1. **решейпим массив**:
   ```python
   A.reshape(A.shape[:-2] + (-1,))
   ```
   - `A.shape` — это `(2, 5, 20, 25)`.
   - `A.shape[:-2]` — это `(2, 5)`, что означает сохранение первых двух измерений.
   - `(-1,)` — это механизм, позволяющий "сжать" оставшиеся два измерения (20 и 25) в одно. Таким образом, итоговая форма после reshape будет `(2, 5, 500)`, где 500 — это произведение 20 и 25 (после сжатия размерности).

   Таким образом, `A.reshape(A.shape[:-2] + (-1,))` преобразует четырёхмерный массив в трёхмерный, где последние два измерения объединены в одно. Это даёт структуру `(2, 5, 500)`.

2. **суммируем по последней оси**:
   ```python
   .sum(axis=-1)
   ```
   После изменения формы массива мы вызываем метод `.sum(axis=-1)`. Это означает, что суммирование происходит по последней оси, то есть по 500-ти значным векторами, которые получились после reshape.

   Таким образом, для каждого элемента с индексами \([i, j]\) мы суммируем все элементы в третьей оси (размерность 500), то есть все элементы с фиксированными индексами \(i\) и \(j\) по последним двум осям исходного массива.

   В результате мы получаем двумерный массив размера `(2, 5)`, где каждый элемент на позиции \([i, j]\) — это сумма всех элементов из исходного массива с фиксированными первыми двумя индексами \(i\) и \(j\).

В выводе получаем двумерный массив с размерами `(2, 5)`, содержащий суммы всех элементов исходного массива для каждого сочетания первых двух индексов.

Таким образом, для массива размером `(2, 5, 20, 25)` после выполнения кода результат будет двумерным массивом, где:
- Строки будут соответствовать значениям из первой и второй размерности массива.
- Столбцы будут соответствовать второй и третьей размерности массива.
- Элементы этого массива — это суммы всех элементов исходного четырёхмерного массива с заданными первыми двумя индексами.


##### Даны одномерные массивы A и B. Элементы массива B принимают значения от 0 до `len(A) - 1`. Требуется прибавить единицу ко всем элементам A, чьи индексы записаны в B. Если индекс встречается в B несколько раз, то надо прибавить единицу для каждого такого вхождения.

In [14]:
A = np.ones(10)
B = np.random.randint(0,len(A),20)
print(A)
print(B)
print(np.bincount(B, minlength=len(A)))
A += np.bincount(B, minlength=len(A))
print(A)

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
[9 7 0 5 7 5 6 3 6 0 9 5 8 0 1 2 9 9 1 5]
[3 2 1 1 0 4 2 2 1 4]
[4. 3. 2. 2. 1. 5. 3. 3. 2. 5.]


Используем функцию np.bincount для подсчёта вхождений индексов: 

   Функция `np.bincount` подсчитывает количество вхождений каждого числа в массиве `B` и возвращает массив, в котором на позиции `i` находится количество вхождений индекса `i` в массив `B`. Параметр `minlength` задаёт минимальную длину результата (в данном случае равную длине массива `A`).

   Например, для массива `B`:
   ```python
   B = [0, 3, 7, 2, 4, 5, 0, 0, 3, 9, 7, 6, 2, 8, 9, 2, 4, 6, 1, 0]
   ```
   функция `np.bincount(B, minlength=len(A))` создаст массив, где на каждом индексе будет подсчитано количество вхождений этого индекса в `B`. Например, для `B` массив будет выглядеть так:
   ```python
   [4, 1, 3, 2, 2, 1, 2, 2, 1, 2]
   ```
   Это означает, что индекс 0 встречается 4 раза, индекс 1 — 1 раз, индекс 2 — 3 раза, индекс 3 — 2 раза и так далее.

**Прибавление подсчитанных значений к массиву `A`**:
   После подсчёта вхождений каждого индекса с помощью `np.bincount`, мы прибавляем полученный массив к массиву `A`. Это приводит к тому, что каждый элемент массива `A` увеличивается на количество раз, которое этот индекс встречается в массиве `B`.

Этот код позволяет эффективно обновить значения в массиве `A` на основе повторений индексов, которые содержатся в массиве `B`, используя функцию `np.bincount` для подсчёта вхождений.

##### Даны одномерный массив A и число n. Вычислите массив B, в котором i-й элемент равен среднему значению элементов с i-го по (i+n-1)-й в массиве A.

In [58]:
def moving_average(Z, n=3) :
    ret = np.cumsum(Z, dtype=float)
    ret[n:] = ret[n:] - ret[:-n]
    return ret[n - 1:] / n
A = np.random.randint(0, 10, 20)
print(A)
print(moving_average(A, n=3))

[6 5 7 6 6 1 7 8 1 1 8 7 8 8 1 5 2 9 2 4]
[6.         6.         6.33333333 4.33333333 4.66666667 5.33333333
 5.33333333 3.33333333 3.33333333 5.33333333 7.66666667 7.66666667
 5.66666667 4.66666667 2.66666667 5.33333333 4.33333333 5.        ]


Входной массив:

```
A = [6, 5, 7, 6, 6, 1, 7, 8, 1, 1, 8, 7, 8, 8, 1, 5, 2, 9, 2, 4]
```

Функция `moving_average(A, n=3)`:

Эта функция вычисляет скользящее среднее для массива `A` с окном шириной `n=3`. Таким образом, для каждого элемента `i` в результирующем массиве будет вычисляться среднее значение трех элементов массива `A`, начиная с индекса `i`.

 1. Накопленная сумма (`np.cumsum(A)`):

Первым шагом в функции идет вычисление накопленной суммы элементов массива `A`:

```python
ret = np.cumsum(A)
```

Результат накопленной суммы:

```
ret = [6, 11, 18, 24, 30, 31, 38, 46, 47, 48, 56, 63, 71, 79, 80, 85, 87, 96, 98, 102]
```

 2. Вычитание старых значений из новых:

Затем, чтобы получить суммы элементов с i-го по (i+n-1)-й, мы выполняем операцию вычитания:

```python
ret[n:] = ret[n:] - ret[:-n]
```

Для `n=3`, вычитание будет выглядеть так:

```
ret[3:] = ret[3:] - ret[:-3]
```

Результат:

```
ret = [6, 11, 18, 18, 24, 30, 38, 46, 47, 48, 56, 63, 71, 79, 80, 85, 87, 96, 98, 102]
```

Теперь у нас есть массив, где каждый элемент (начиная с индекса 3) представляет собой сумму трех последовательных элементов из массива `A`.

 3. Разделение на `n`:

Наконец, для получения скользящего среднего, мы делим каждый элемент накопленной суммы на `n=3`:

```python
ret[n - 1:] / n
```

Результат:

```
[6.0, 6.0, 6.33333333, 4.33333333, 4.66666667, 5.33333333, 5.33333333, 3.33333333, 3.33333333, 5.33333333, 7.66666667, 7.66666667, 5.66666667, 4.66666667, 2.66666667, 5.33333333, 4.33333333, 5.0]
```

Это и есть результат выполнения функции — массив, в котором каждый элемент равен среднему значению трех последовательных элементов массива `A`.

 Подробное объяснение вывода:

- Первый элемент результата равен `6.0`, потому что первые три элемента массива `A` — это `[6, 5, 7]`, и их среднее значение равно `(6 + 5 + 7) / 3 = 6.0`.
- Второй элемент тоже равен `6.0`, потому что сумма элементов на индексах 1, 2 и 3 в `A` — это `[5, 7, 6]`, среднее равно `(5 + 7 + 6) / 3 = 6.0`.
- Третий элемент равен `6.33`, потому что сумма элементов на индексах 2, 3 и 4 — это `[7, 6, 6]`, среднее равно `(7 + 6 + 6) / 3 = 6.333`.
- И так далее.

Этот процесс повторяется для всех элементов массива, начиная с индекса 2, где для каждого индекса мы берем три элемента массива и делаем их среднее.


### Задания для самостоятельного решения

1. Развернуть одномерный массив (сделать так, чтобы его элементы шли в обратном порядке).
2. Найти максимальный нечетный элемент в массиве.
3. Замените все нечетные элементы массива на ваше любимое число.
4. Создайте массив первых n нечетных чисел, записанных в порядке убывания. Например, если `n=5`, то ответом будет `array([9, 7, 5, 3, 1])`. *Функции, которые могут пригодиться при решении: `.arange()`*
5. Вычислите самое близкое и самое дальнее числа к данному в рассматриваемом массиве чисел. Например, если на вход поступают массив `array([0, 1, 2, 3, 4])` и число 1.33, то ответом будет `(1, 4)`. _Функции, которые могут пригодиться при решении: `.abs()`, `.argmax()`, `.argmin()`_
6. Вычисляющую первообразную заданного полинома (в качестве константы возьмите ваше любимое число). Например, если на вход поступает массив коэффициентов `array([4, 6, 0, 1])`, что соответствует полиному $4x^3 + 6x^2 + 1$, на выходе получается массив коэффициентов `array([1, 2, 0, 1, -2])`, соответствующий полиному $x^4 + 2x^3 + x - 2$. _Функции, которые могут пригодиться при решении: `.append()`_
7. Пользуясь пунктом 6, посчитайте первую производную для заданного полинома в заданной точке.